### ISSUE TITLE
Allow Init Checker to look within anonymous classes

### ISSUE THREAD

**Issue by bishabosha** [lampepfl/dotty#10549]
```scala
class Wrap {
  class E
  object E {
    final val A = new E {}
    val $values = Array(A)
  }
}
```
produces warning with `-Ycheck-init`
```scala
-- Warning: initcheck.scala:5:23 ----------------------------------
5 |    val $values = Array(A)
  |                        ^
  |Promoting the value under initialization to fully-initialized. Calling trace:
  | -> val $values = Array(A)	[ initcheck.scala:5 ]
1 warning found
```
when it could look inside to prove it is safe.

One example that is not safe and will lead to NPE in initialisation of object `E`:
```scala
class Wrap {                                                                                                     
  def qux[T](e: E[T]) = e.foo

  abstract class E[+T] { def foo: T }
  object E {
    final val A: E[Nothing] = new E { def foo = ref }
    val ref = qux(A)
  }
}
```
as pointed out in https://github.com/lampepfl/dotty/pull/9665#issuecomment-682536402

if `ref` is made lazy then there is no warning on `Array(A)` and object `E` can be constructed correctly.

**Comment by propensive** [comment]
This issue is referenced in a comment here: https://github.com/lampepfl/dotty/blob/4a6ec60f13be412e11bb81b3c3028cc47dc1c2ec/compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala#L122-L127

Given that this issue is closed, it might be possible to address the comment now.

**Comment by liufengyun** [comment]
This issue will be fixed by #11533. @natsukagami Could you add it as a test:

- `tests/init/pos/i10549a.scala`
- `tests/init/neg/i10549b.scala`
